### PR TITLE
Fetch ad-hoc query metadata on the BE

### LINKED ISSUE #43628: BE - Add `/api/dataset/query_metadata` endpoint
Similar to the endpoints introduced here https://github.com/metabase/metabase/issues/43330

### LINKED ISSUE #43629: FE - Use `/api/dataset/query_metadata` endpoint

### PR DESCRIPTION
**Pull request by ranquild**
Epic https://github.com/metabase/metabase/issues/40898
Fixes https://github.com/metabase/metabase/issues/43628
Fixes https://github.com/metabase/metabase/issues/43629
Unblocks https://github.com/metabase/metabase/issues/43630

Makes all query metadata to be managed by RTK without entity framework caching. This became possible with recent `query_metadata` endpoints that fetch all metadata at once. The only missing piece was `/api/dataset/query_metadata` for adhoc queries and removing entity framework caching, which this PR does.

`EditSandboxingModal` created invalid adhoc queries (without database id), so I had to fix it otherwise the new endpoint won't work.

How to verify:
- CI is green

### CODE REVIEW

File: frontend/src/metabase-types/guards/card.ts [hunk 1]
```diff
@@ -1,5 +1,5 @@
-import type { Card, UnsavedCard, XrayCard } from "metabase-types/api";
+import type { Card, UnsavedCard } from "metabase-types/api";
 
-export function isSavedCard(card: Card | UnsavedCard | XrayCard): card is Card {
```
Comment (ranquild) [inline frontend/src/metabase-types/guards/card.ts:3]: Metadata for xray cards is no longer fetched on the FE

File: frontend/src/metabase/redux/metadata.js [hunk 2]
```diff
@@ -250,32 +248,3 @@ export const fetchRealDatabasesWithMetadata = createThunkAction(
     };
   },
 );
-
-export const loadMetadataForDependentItems =
```
Comment (ranquild) [inline frontend/src/metabase/redux/metadata.js:254]: Finally we're removing this

File: enterprise/frontend/src/metabase-enterprise/sandboxes/utils.ts [hunk 3]
```diff
@@ -10,9 +10,10 @@ export const getPolicyKeyFromParams = ({
 export const getPolicyKey = (policy: GroupTableAccessPolicy) =>
   `${policy.group_id}:${policy.table_id}`;
 
-export const getRawDataQuestionForTable = (tableId: TableId) => ({
+export const getRawDataQuestionForTable = (table: Table) => ({
   dataset_query: {
     type: "query",
-    query: { "source-table": tableId },
+    database: table?.db_id,
+    query: { "source-table": table?.id },
```
Comment (kamilmielnik) [inline enterprise/frontend/src/metabase-enterprise/sandboxes/utils.ts:17]:
```suggestion
    database: table.db_id,
    query: { "source-table": table.id },
```

File: frontend/src/metabase/api/dataset.ts [hunk 4]
```diff
@@ -1,9 +1,23 @@
...
+  DatasetQuery,
+  CardQueryMetadata,
+} from "metabase-types/api";
 
 import { Api } from "./api";
+import { provideAdhocQueryMetadataTags } from "./tags";
 
 export const datasetApi = Api.injectEndpoints({
   endpoints: builder => ({
+    getAdhocQueryMetadata: builder.query<CardQueryMetadata, DatasetQuery>({
```
Comment (kamilmielnik) [inline frontend/src/metabase/api/dataset.ts:12]: `CardQueryMetadata` uses `Table` type which is currently inaccurate for card-based _tables_.
We might want to improve the typing there in the near future.

image

**Review by kamilmielnik — APPROVED** [review]
LGTM :+1:

File: frontend/src/metabase/questions/actions.ts [hunk 5]
```diff
@@ -1,103 +1,23 @@
...
-      prevDependencies,
-      nextDependencies,
-    );
-    if (newDependencies.length > 0) {
-      const mergedDependencies = [...prevDependencies, ...newDependencies];
-      await dispatch(loadMetadataForDependentItems(newDependencies, options));
-      await dispatch(
-        loadMetadata(getDependencies, mergedDependencies, options),
-      );
+  (card: Card | UnsavedCard) => async (dispatch: Dispatch) => {
```
Comment (uladzimirdev) [inline frontend/src/metabase/questions/actions.ts:17]:
```suggestion
  (card: Card | UnsavedCard) => (dispatch: Dispatch) => {
```

File: frontend/src/metabase-types/guards/card.ts [hunk 6]
```diff
@@ -1,5 +1,5 @@
...
-  return "id" in card && typeof card.id === "number";
+export function isSavedCard(card: Card | UnsavedCard): card is Card {
+  return "id" in card && card.id != null;
```
Comment (uladzimirdev) [inline frontend/src/metabase-types/guards/card.ts:4]:
```suggestion
  return card.id != null;
```

**Review by uladzimirdev — APPROVED** [review]
LGTM

**Pull request merged**